Open edX Teak release notes - incomplete first cut (#1127)
openedx/docs.openedx.org: source/community/release_notes/teak/dev_op_release_notes.rst

           mismatches in behavior between the two environments.
         - Span-based telemetry (New Relic, Datadog, etc.) can be used to track rates of mismatches and break them down by course ID and type. See 
           ``set_custom_attribute`` calls starting with ``codejail``. in 
           `safe_exec.py <https://github.com/openedx/edx-platform/blob/release/teak.master/xmodule/capa/safe_exec/safe_exec.py>`_  
           for available attributes. The local-only, remote-only and local/remote darklaunch calls all have different span names as well, e.g. 
           ``safe_exec.remote_exec_darklaunch``.
         - Use ``CODEJAIL_DARKLAUNCH_EMSG_NORMALIZERS`` to normalize away spurious mismatches between the environments. (Not all mismatches can be readily ignored, such 
           as ordering differences in sets.)

      - Once behavior and performance differences are resolved, remove ``ENABLE_CODEJAIL_DARKLAUNCH`` and set ``ENABLE_CODEJAIL_REST_SERVICE`` to true. This will complete 
        the migration, and codejail executions will only be performed on the remote service.

>          

Reviewer: @cmltaWt0 should we put an operator note about Credentials/Badging?

Author: @cmltaWt0 You can mention whatever you want :) I'm just the janitor putting it all together :) You can put the content you want as comments here inline, or add them to [the wiki](https://docs.openedx.org/en/latest/community/release_notes/teak/dev_op_release_notes.html).

Reviewer 2: Ahh. Sure, let me add it to the wiki. Thanks!

Reviewer: @cmltaWt0 this is the link to the wiki, please let us know when you've added an entry. https://openedx.atlassian.net/wiki/spaces/COMM/pages/4570710024/Teak+-+Operator+Dev+Notes

Reviewer: @cmltaWt0 ping on this.

      - You should be able to identify these exclusions by looking for lines like ``/home/sandbox/.config/ wrix``, although the exact parent directory may vary. Other
        temporary directories may have been allowed instead, such as ``/tmp``. Any such write permission to a global directory is inadvisable, since it reduces the 
        ability of codejail to perform effective sandboxing. Removing these lines in Teak will (appropriately) reduce the permissions of sandboxed code. They should 
        not be removed before Teak, however, as this will cause matplotlib to fail to load.
      - Operators who have not previously needed to support matplotlib in instructor or learner code may not have these 
        exclusions in their AppArmor configurations.
        If this is your situation, no action is required.
      - Removing these lines may cause other, unanticipated failures in sandboxed code. Monitor your codejail logs and 
        failure rates when deploying this change.
- **All MFEs have been upgraded to React 18**.  If you have forked an MFE and want to pull in the upstream changes from Teak, you will need to ensure all dependencies work with React 18.

> 

Reviewer: ```suggestion

- MFE slots now follow a versioned naming convention, for example, ``footer_slot`` is now ``org.openedx.frontend.layout.footer.v1``.

  - Aliases are present for existing slots, and will be throughout the lifecycle of the available slot. Any version bump to a slot will have an associated DEPR ticket.
  - See the `slot renaming ADR <https://github.com/openedx/frontend-plugin-framework/blob/master/docs/decisions/0003-slot-naming-and-life-cycle.rst>`_ for more details on the new naming scheme. All new slots will use this naming scheme, and developers making new slots should pay attention to the ADR.
```

Other Operator Changes
======================

- In LMS and CMS, Celery now uses task protocol 2.
   - *Action*: Any operator using custom Celery tooling should ensure it is compatible with protocol 2. For other operators, no action is required.
   - *Background*: Celery 4.0 `switched how task messages are structured <https://docs.celeryq.dev/en/stable/history/whatsnew-4.0.html#new-protocol-highlights>`_ and
     the new message format is called protocol 2. The version of Celery we currently use (anything >=4.0) can create and consume both protocol versions and it should
     be safe to switch between them with zero downtime.

      - By default, Celery 4.0 and higher produce messages in this format, and Celery 3.1.25 and higher can read messages in this format.
      - edx-platform was pinned to protocol 1 `during the upgrade <https://github.com/openedx/edx-platform/pull/24916>`_ to Celery 4, presumably as a precaution. 
        This change is the long-delayed unpinning of the protocol version so that Celery can use its default version.

>       - Operators can still override the protocol version using the Django setting `CELERY_TASK_PROTOCOL` although there is no guarantee that protocol 1 compatibility

Reviewer: For code things (file paths, settings, flags) it'd be great to render these in code formatting, with two backticks:

```suggestion
      - Operators can still override the protocol version using the Django setting ``CELERY_TASK_PROTOCOL`` although there is no guarantee that protocol 1 compatibility
```

Not going to flag all of them but if you could go through and find them that'd be great (that includes things like import matlib, ./tmp/, etc below

  <https://github.com/openedx/brand-openedx/issues/23>`_, and follow the
  :ref:`Ulmo Design Tokens` page for more detail. Operators will be able to try
  out Design Tokens using the Teak Design Tokens branches (link TBD)

> - `[DEPR]: block_structure.storage_backing_for_cache in edx-platform <https://github.com/openedx/public-engineering/issues/32>`_ This is a simplification to how course content is cached. It should be invisible to all end users.

Reviewer: @feanil could you or someone from DEPR review the DEPR tickets in the release notes and let us know if there are others to be added? I find it hard to believe there's only been 3 since late October.

- `[DEPR]: django-oauth2-provider (DOP) related tables <https://github.com/openedx/public-engineering/issues/82>`_

   - For LMS and CMS, there is a new script to 
     `clean up old DOP-related authentication tables. <https://github.com/edx/configuration/blob/master/util/drop_dop_tables/drop_dop_tables.py>`_
   - If you have an old installation of the Open edX platform (Palm or later), you may have many outdated/unused 
     authentication-related tables that can lead to confusion when looking at the database.
   - The script is not related to Teak other than it now being available, and should be ok to run on any installation using Palm, Redwood, or Sumac.
- `[DEPR]: Support for footer replacement via npm installing forked footers <https://github.com/openedx/frontend-component-footer/issues/459>`_
   - It is possible to work around this breaking change by also exporting your forked ``Footer`` component as ``FooterSlot`` and your forked ``StudioFooter`` component as ``StudioFooterSlot``

> 

Reviewer: @kdmccormick - I think this is where to add an operator note about the `legacy_studio*` flags.

  <https://github.com/openedx/brand-openedx/issues/23>`_, and follow the
  :ref:`Ulmo Design Tokens` page for more detail. Operators will be able to try

>   out Design Tokens using the Teak Design Tokens branches (link TBD)

Reviewer: ```suggestion
  out Design Tokens using the Teak Design Tokens branches. See the `Teak Design Tokens wiki page <https://openedx.atlassian.net/wiki/spaces/BPL/pages/5050499077/Using+Teak+Design+Tokens+branches>`_ for more details..
```
